File: problems/260_single_number_iii/Solution.hs
-- Leetcode 260. Single Number III
module Solution where

import Data.Bits (xor, (.&.))
import Data.List (foldl')

singleNumberFold :: [Int] -> [Int]
singleNumberFold nums = [a, b]
  where
    -- XOR all numbers to get a ^ b
    xorSum = foldl' xor 0 nums

    -- Isolate the lowest set bit of xorSum.
    -- In two's complement, -x = ~x + 1, so x .&. (-x) isolates the lowest set bit.
    mask = xorSum .&. (-xorSum)

    -- Fold over the list, accumulating (groupA, groupB) XOR results.
    -- Initial accumulator is (0, 0).
    (a, b) = foldl' step (0, 0) nums

    -- Helper: update the appropriate group based on the mask.
    step (accA, accB) x
      -- If x has the distinguishing bit set, XOR it into the first group.
      | (x .&. mask) /= 0 = (accA `xor` x, accB)
      -- Otherwise, XOR it into the second group.
      | otherwise         = (accA, accB `xor` x)

File: problems/260_single_number_iii/solution.cpp
// Leetcode 260. Single Number III

// Algorithm:
//   1. XOR all numbers -> xor_sum = a ^ b (the XOR of the two unique numbers)
//   2. Find the lowest set bit in xor_sum. This bit is 1 in either a or b, but
//      not both.
//   3. Partition numbers into two groups: those with that bit set, and those
//      without. We are certain that two numbers are in different group
//   4. XOR within each group to isolate a and b separately.
#include <numeric>
#include <vector>

class Solution {
 public:
  std::vector<int> singleNumber(const std::vector<int>& nums) {
    unsigned int xor_sum = 0;
    for (const int num : nums) {
      xor_sum ^= num;
    }

    const unsigned int lowest_set_bit = xor_sum & ~(xor_sum - 1);

    int first_single = 0;
    int second_single = 0;

    for (const int num : nums) {
      if ((num & lowest_set_bit) != 0) {
        first_single ^= num;
      } else {
        second_single ^= num;
      }
    }

    return {first_single, second_single};
  }
};
